File: AutoRest/Generators/NodeJS/NodeJS.Tests/AcceptanceTests/requiredOptionalTests.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

'use strict';

var should = require('should');
var http = require('http');
var util = require('util');
var assert = require('assert');
var msRest = require('ms-rest');

var reqOptClient = require('../Expected/AcceptanceTests/RequiredOptional/autoRestRequiredOptionalTestService');

var dummyToken = 'dummy12321343423';
var credentials = new msRest.TokenCredentials(dummyToken);

var clientOptions = {};
var baseUri = 'http://localhost:3000';


describe('nodejs', function () {

  describe('Swagger Required Optional BAT', function () {

    describe('Basic Required Optional Operations', function () {
      var testClient = new reqOptClient('', '', baseUri, clientOptions);

      it('should throw error on null path parameter', function (done) {
        testClient.implicit.getRequiredPath(null, function (error, result) {
          should.exist(error);
          error.message.should.containEql('pathParameter cannot be null or undefined and it must be of type string.');
          done();
        });
      });

      it('should accept null values for query parameters', function (done) {
        testClient.implicit.putOptionalQuery(null, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });

      it('should accept null values for optional header parameters', function (done) {
        testClient.implicit.putOptionalHeader(null, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });

      it('should accept null values for optional body parameters', function (done) {
        testClient.implicit.putOptionalBody(null, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });

      it('should throw error on null values for required integer parameters', function (done) {
        testClient.explicit.postRequiredIntegerParameter(null, function (error, result) {
          should.exist(error);
          error.message.should.containEql('bodyParameter cannot be null or undefined and it must be of type number.');
          done();
        });
      });

      it('should accept null values for optional integer parameters', function (done) {
        testClient.explicit.postOptionalIntegerParameter(null, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });

      it('should throw error on null values for required integer properties', function (done) {
        testClient.explicit.postRequiredIntegerProperty({ 'value': null }, function (error, result) {
          should.exist(error);
          error.message.should.containEql('value');
          error.message.should.containEql('cannot be null or undefined');
          done();
        });
      });

      it('should accept null values for optional integer properties', function (done) {
        testClient.explicit.postOptionalIntegerProperty({ 'value': null }, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });

      it('should throw error on null values for required integer header', function (done) {
        testClient.explicit.postRequiredIntegerHeader(null, function (error, result) {
          should.exist(error);
          error.message.should.containEql('headerParameter cannot be null or undefined and it must be of type number.');
          done();
        });
      });

      it('should accept null values for optional integer header', function (done) {
        testClient.explicit.postOptionalIntegerHeader(null, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });

      it('should throw error on null values for required string parameters', function (done) {
        testClient.explicit.postRequiredStringParameter(null, function (error, result) {
          should.exist(error);
          error.message.should.containEql('bodyParameter cannot be null or undefined and it must be of type string.');
          done();
        });
      });

      it('should accept null values for optional string parameters', function (done) {
        testClient.explicit.postOptionalStringParameter(null, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });

      it('should throw error on null values for required string properties', function (done) {
        testClient.explicit.postRequiredStringProperty({ 'value': null }, function (error, result) {
          should.exist(error);
          error.message.should.containEql('value');
          error.message.should.containEql('cannot be null or undefined');
          done();
        });
      });

      it('should accept null values for optional string properties', function (done) {
        testClient.explicit.postOptionalStringProperty({ 'value': null }, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });

      it('should throw error on null values for required string header', function (done) {
        testClient.explicit.postRequiredStringHeader(null, function (error, result) {
          should.exist(error);
          error.message.should.containEql('headerParameter cannot be null or undefined and it must be of type string.');
          done();
        });
      });

      it('should accept null values for optional string header', function (done) {
        testClient.explicit.postOptionalStringHeader(null, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });


      it('should throw error on null values for required class parameters', function (done) {
        testClient.explicit.postRequiredClassParameter(null, function (error, result) {
          should.exist(error);
          error.message.should.containEql('bodyParameter cannot be null or undefined.');
          done();
        });
      });

      it('should accept null values for optional class parameters', function (done) {
        testClient.explicit.postOptionalClassParameter(null, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });

      it('should throw error on null values for required class properties', function (done) {
        testClient.explicit.postRequiredClassProperty({ 'value': null }, function (error, result) {
          should.exist(error);
          error.message.should.containEql('value');
          error.message.should.containEql('cannot be null or undefined');
          done();
        });
      });

      it('should accept null values for optional class properties', function (done) {
        testClient.explicit.postOptionalClassProperty({ 'value': null }, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });


      it('should throw error on null values for required array parameters', function (done) {
        testClient.explicit.postRequiredArrayParameter(null, function (error, result) {
          should.exist(error);
          error.message.should.containEql('bodyParameter cannot be null or undefined and it must be of type array.');
          done();
        });
      });

      it('should accept null values for optional array parameters', function (done) {
        testClient.explicit.postOptionalArrayParameter(null, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });

      it('should throw error on null values for required array properties', function (done) {
        testClient.explicit.postRequiredArrayProperty({ 'value': null }, function (error, result) {
          should.exist(error);
          error.message.should.containEql('value');
          error.message.should.containEql('cannot be null or undefined');
          done();
        });
      });

      it('should accept null values for optional array properties', function (done) {
        testClient.explicit.postOptionalArrayProperty({ 'value': null }, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });

      it('should throw error on null values for required array header', function (done) {
        testClient.explicit.postRequiredArrayHeader(null, function (error, result) {
          should.exist(error);
          error.message.should.containEql('headerParameter cannot be null or undefined and it must be of type array.');
          done();
        });
      });

      it('should accept null values for optional array header', function (done) {
        testClient.explicit.postOptionalArrayHeader(null, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });

      it('should throw error on null global property in path', function (done) {
        testClient.requiredGlobalPath = null;
        testClient.implicit.getRequiredGlobalPath(function (error, result) {
          should.exist(error);
          error.message.should.containEql('this.client.requiredGlobalPath cannot be null or undefined and it must be of type string.');
          done();
        });
      });

      it('should throw error on null global property in query', function (done) {
        testClient.requiredGlobalQuery = null;
        testClient.implicit.getRequiredGlobalQuery(function (error, result) {
          should.exist(error);
          error.message.should.containEql('this.client.requiredGlobalQuery cannot be null or undefined and it must be of type string.');
          done();
        });
      });

      it('should accept null values for optional global property in query', function (done) {
        testClient.implicit.getOptionalGlobalQuery(null, function (error, result) {
          should.not.exist(error);
          result.response.statusCode.should.equal(200);
          done();
        });
      });

    });
  });
});

File: AutoRest/Generators/Python/Python.Tests/AcceptanceTests/required_optional_tests.py
import unittest
import subprocess
import sys
import isodate
from datetime import date, datetime, timedelta
from os.path import dirname, pardir, join, realpath, sep, pardir

cwd = dirname(realpath(__file__))
root = realpath(join(cwd , pardir, pardir, pardir, pardir, pardir))
sys.path.append(join(root, "ClientRuntimes" , "Python", "msrest"))

tests = realpath(join(cwd, pardir, "Expected", "AcceptanceTests"))
sys.path.append(join(tests, "RequiredOptional"))

from msrest.exceptions import DeserializationError, SerializationError

from auto_rest_required_optional_test_service import (
    AutoRestRequiredOptionalTestService,
    AutoRestRequiredOptionalTestServiceConfiguration)

from auto_rest_required_optional_test_service.models import StringWrapper, ArrayWrapper, ClassWrapper


class RequiredOptionalTests(unittest.TestCase):

    @classmethod
    def setUpClass(cls):

        config = AutoRestRequiredOptionalTestServiceConfiguration(
            "required_path",
            "required_query",
            base_url="http://localhost:3000")

        config.log_level = 10
        cls.client = AutoRestRequiredOptionalTestService(config)
        return super(RequiredOptionalTests, cls).setUpClass()

    def test_required_optional(self):

        self.client.config.required_global_path = "required_path"
        self.client.config.required_global_query = "required_query"

        self.client.implicit.put_optional_query(None)
        self.client.implicit.put_optional_body(None)
        self.client.implicit.put_optional_header(None)

        self.client.implicit.get_optional_global_query(None)

        self.client.explicit.post_optional_integer_parameter(None)
        self.client.explicit.post_optional_integer_property(None)
        self.client.explicit.post_optional_integer_header(None)

        self.client.explicit.post_optional_string_parameter(None)
        self.client.explicit.post_optional_string_property(None)
        self.client.explicit.post_optional_string_header(None)

        self.client.explicit.post_optional_class_parameter(None)
        self.client.explicit.post_optional_class_property(None)

        self.client.explicit.post_optional_array_parameter(None)
        self.client.explicit.post_optional_array_property(None)
        self.client.explicit.post_optional_array_header(None)

    def test_required_optional_negative(self):

        self.client.config.required_global_path = None
        self.client.config.required_global_query = None

        with self.assertRaises(ValueError):
            self.client.implicit.get_required_path(None)

        with self.assertRaises(ValueError):
            self.client.explicit.post_required_string_header(None)

        # TODO need check body parameter
        #with self.assertRaises(ValueError):
        #    self.client.explicit.post_required_string_parameter(None)

        with self.assertRaises(SerializationError):
            self.client.explicit.post_required_string_property(None)

        # TODO, is ValueError or TypeError, should check the None first
        with self.assertRaises(TypeError):
            self.client.explicit.post_required_array_header(None)

        # TODO, need check body parameter
        #with self.assertRaises(TypeError):
        #    self.client.explicit.post_required_array_parameter(None)

        with self.assertRaises(SerializationError):
            self.client.explicit.post_required_array_property(None)

        # TODO need check body parameter
        #with self.assertRaises(ValueError):
        #    self.client.explicit.post_required_class_parameter(None)

        cw = ClassWrapper(value = None)
        with self.assertRaises(SerializationError):
            self.client.explicit.post_required_class_property(cw)

        with self.assertRaises(ValueError):
            self.client.implicit.get_required_global_path()

        with self.assertRaises(ValueError):
            self.client.implicit.get_required_global_query()


if __name__ == '__main__':
    
    
    unittest.main()